File: src/main/java/hu/finominfo/scheduler/people/People.java
package hu.finominfo.scheduler.people;

import hu.finominfo.scheduler.common.Globals;
import hu.finominfo.scheduler.util.KeyValueStore;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.sql.SQLException;
import java.util.*;

/**
 * Created by kalman.kovacs@globessey.local on 2017.12.18.
 */
public class People {
    private final Map<String, Person> people = new HashMap<>();
    private final Map<Integer, Set<String>> hated = new HashMap<>();

    public People() throws IOException, SQLException {
        String content = new String(Files.readAllBytes(Paths.get(Globals.getInstance().getConfigFile())), "UTF-8");
        String lines[] = content.split("\\r?\\n");
        for (int i = 0; i < 31 + 1; i++) {
            hated.put(i, new HashSet<>());
        }
        KeyValueStore keyValueStore = new KeyValueStore();
        List<String> names = keyValueStore.getNames();
        keyValueStore.close();
        List<String> keywords = Arrays
                .asList(new String[] { "nofo", "hend", "hweek", "hmon", "htue", "hwen", "hthu", "hfri", "wtue" });
        Arrays.asList(lines).stream().forEach(line -> {
            Person person = null;
            Person newPerson = null;
            for (String expression : line.trim().split(" ")) {
                final String trimmedExpression = expression.trim();
                if (trimmedExpression.isEmpty()) {
                    continue;
                }
                if (person == null) {
                    if (newPerson != null) {
                        if(trimmedExpression.equalsIgnoreCase("newperson")) {
                            person = newPerson;
                            people.put(person.getName(), person);
                        } else {
                            throw new RuntimeException("New person (" + newPerson.getName() + ") allowed only with newperson command" );
                        }
                    } else {
                        String replaced = trimmedExpression.replaceAll("_", " ");
                        if (names.contains(replaced)) {
                            person = new Person(replaced);
                            people.put(replaced, person);
                        } else {
                            newPerson = new Person(replaced);
                        }
                    }
                } else if (keywords.contains(trimmedExpression.toLowerCase())) {
                    switch (trimmedExpression.toLowerCase()) {
                        case "nofo":
                            for (int day : person.getTypes().keySet()) {
                                person.setType(day, Type.BO);
                            }
                            person.setNofo(true);
                            break;
                        case "hend":
                            person.setHatesWeekends(true);
                            break;
                        case "hweek":
                            person.setHatesWeekdays(true);
                            break;
                        case "hmon":
                            person.setHatesMondays(true);
                            break;
                        case "htue":
                            person.setHatesTuesdays(true);
                            break;
                        case "hwen":
                            person.setHatesWednesdays(true);
                            break;
                        case "hthu":
                            person.setHatesThursdays(true);
                            break;
                        case "hfri":
                            person.setHatesFridays(true);
                            break;
                        case "wtue":
                            person.setWantedTuesdays(true);
                            break;
                    }
                } else if (trimmedExpression.length() >= 1){
                    String remaining = trimmedExpression.substring(1);
                    final List<Integer> days = new ArrayList<Integer>();
                    if (remaining.contains("-")) {
                        String[] split = remaining.split("-");
                        int start = Integer.valueOf(split[0]);
                        int end = Integer.valueOf(split[1]);
                        while (start <= end) {
                            days.add(start);
                            start++;
                        }
                    } else {
                        days.add(Integer.valueOf(remaining));
                    }
                    switch (trimmedExpression.charAt(0)) {
                        case 'w':
                            setWantedDays(person, days);
                            break;
                        case 'h':
                            setHatedDays(person, days);
                            break;
                        case 'f':
                            for (int day : days)
                                person.setType(day, Type.FO);
                            //setWantedDays(person, days);
                            break;
                        case 'b':
                            for (int day : days)
                                person.setType(day, Type.BO);
                            //setWantedDays(person, days);
                            break;
                        case '+':
                            person.getManualDayDifference().getAndSet(days.get(0));
                            break;
                        case '-':
                            person.getManualDayDifference().getAndSet(-(days.get(0)));
                            break;
                    }
                }
            }
        });
    }

    private static void setHatedDays(Person person, List<Integer> days) {
        person.getHatedDays().addAll(days);
        for (int day : days) {
            if (person.getWantedDays().contains(day)) {
                throw new RuntimeException(
                        person.getName() + " wants and hates the same day: " + day);
            }
        }
    }

    private static void setWantedDays(Person person, List<Integer> days) {
        person.getWantedDays().addAll(days);
        for (int day : days) {
            if (person.getHatedDays().contains(day)) {
                throw new RuntimeException(
                        person.getName() + " wants and hates the same day: " + day);
            }
        }
    }

    public Map<String, Person> getPeople() {
        return people;
    }
}

File: scheduler_python/people/people.py
# scheduler_python/people/people.py

import sys
import os
import re
import codecs
from ..people.person import Person
from ..people.type import Type
from ..common.globals import Globals
from ..util.key_value_store import KeyValueStore

class People:
    """
    Python version of the Java People class.
    It reads lines from config.csv, interprets each line (with same logic).
    """
    def __init__(self):
        self.people = {}
        self.hated = {}
        for i in range(32):
            self.hated[i] = set()
        g = Globals()
        config_file = g.get_config_file()

        # In Java, it reads the entire file into 'content' and splits lines.
        # We'll do similarly in Python:
        with codecs.open(config_file, "r", encoding="utf-8") as f:
            lines = f.read().splitlines()

        # We get all distinct names from KeyValueStore in Java as well
        # because People tries to see if the name is recognized from DB or "newperson".
        kv = KeyValueStore()
        known_names = kv.get_names()
        kv.close()
        keywords = [
            "nofo","hend","hweek","hmon","htue","hwen","hthu","hfri","wtue"
        ]

        new_person_must_be_tagged = False
        new_person_obj = None

        # The Java code logic is: for each line, split on spaces, parse them.
        for line in lines:
            line = line.strip()
            if not line:
                continue
            expressions = line.split()
            person = None
            # We replicate the logic from People.java
            for expr in expressions:
                expr = expr.strip()
                if not expr:
                    continue

                # If person is still None, we treat expr as the name (or 'newperson'):
                if person is None:
                    if new_person_obj is not None:
                        # Means we got a newPerson line
                        if expr.lower() == "newperson":
                            # finalize new_person
                            person = new_person_obj
                            self.people[person.get_name()] = person
                            new_person_obj = None
                        else:
                            raise RuntimeError("New person declared but 'newperson' keyword not found.")
                    else:
                        # Check if name is known or brand new:
                        replaced = expr.replace("_", " ")
                        if replaced in known_names:
                            # Just create it as a known Person
                            person = Person(replaced)
                            self.people[replaced] = person
                        else:
                            # The Java code logic: if not found among known names, 
                            # it tries to see if the next token is "newperson"? 
                            # We replicate that approach: let's store it in new_person_obj
                            new_person_obj = Person(replaced)
                else:
                    # if expr is in the keywords:
                    if expr.lower() in keywords:
                        if expr.lower() == "nofo":
                            for d in person.types.keys():
                                person.set_type(d, Type.BO)
                            person.set_nofo(True)
                        elif expr.lower() == "hend":
                            person.set_hates_weekends(True)
                        elif expr.lower() == "hweek":
                            person.set_hates_weekdays(True)
                        elif expr.lower() == "hmon":
                            person.set_hates_mondays(True)
                        elif expr.lower() == "htue":
                            person.set_hates_tuesdays(True)
                        elif expr.lower() == "hwen":
                            person.set_hates_wednesdays(True)
                        elif expr.lower() == "hthu":
                            person.set_hates_thursdays(True)
                        elif expr.lower() == "hfri":
                            person.set_hates_fridays(True)
                        elif expr.lower() == "wtue":
                            person.set_wanted_tuesdays(True)
                    else:
                        # not a keyword, parse further
                        c = expr[0]  # first char
                        remaining = expr[1:]
                        # parse days
                        days = []
                        if "-" in remaining:
                            splitted = remaining.split("-")
                            start = int(splitted[0])
                            end = int(splitted[1])
                            for dd in range(start, end+1):
                                days.append(dd)
                        else:
                            days.append(int(remaining))

                        if c == 'u':
                            # set numOfWantedHolidays
                            person.set_num_of_wanted_holidays(days[0])
                        elif c == 'p':
                            # set numOfWantedFridays
                            person.set_num_of_wanted_fridays(days[0])
                        elif c == 's':
                            # set numOfWantedSaturdays
                            person.set_num_of_wanted_saturdays(days[0])
                        elif c == 'v':
                            # set numOfWantedSundays
                            person.set_num_of_wanted_sundays(days[0])
                        elif c == 'w':
                            # wanted days
                            for d in days:
                                person.get_wanted_days().add(d)
                                if d in person.get_hated_days():
                                    raise RuntimeError(f"{person.get_name()} wants and hates day {d} at once")
                        elif c == 'h':
                            # hated days
                            for d in days:
                                person.get_hated_days().append(d)
                                if d in person.get_wanted_days():
                                    raise RuntimeError(f"{person.get_name()} wants and hates day {d} at once")
                        elif c == 'f':
                            # set day type FO
                            for d in days:
                                person.set_type(d, Type.FO)
                        elif c == 'b':
                            # set day type BO
                            for d in days:
                                person.set_type(d, Type.BO)
                        elif c == '+':
                            # manualDayDifference
                            person.set_manual_day_difference(days[0])
                        elif c == '-':
                            # negative manualDayDifference
                            person.set_manual_day_difference(-days[0])
                        else:
                            pass
                            # The Java code doesn't handle other letters differently, so we ignore.

        # If at the end we still have new_person_obj not None, then user forgot "newperson" keyword
        # But let's just ignore or raise. The original Java logic would raise.
        if new_person_obj is not None:
            raise RuntimeError("Found a name that is not in DB but did not get 'newperson' tag in line, can't finalize person")

    def get_people(self):
        return self.people
